app/models/framework_response.rb

```diff
@@ -50,31 +43,27 @@ def value=(raw_value)
     end
   end
 
-private
-
-  def set_responded_value
-    self.responded = true
+  def rebuild_flags
+    self.framework_flags = framework_question.framework_flags.select { |flag| option_selected?(flag.question_value) }
   end
 
-  def build_flags
-    return [] unless framework_question.framework_flags.any?
+  def self.requires_value?(value, record)
+    return false if value.present? || !record.framework_question.required
 
-    framework_question.framework_flags.each_with_object([]) do |flag, arr|
-      if option_selected?(flag.question_value)
-        arr << flag
-      end
-    end
-  end
+    return true if record.parent.blank?
 
-  def clear_dependent_values_and_flags!(old_value)
-    return unless old_value != value
+    record.parent.option_selected?(record.framework_question.dependent_value)
+  end
 
-    dependent_ids = dependents.includes(:framework_question).reject { |dependent| option_selected?(dependent.framework_question.dependent_value) }
+  def self.clear_dependent_values_and_flags!(responses_to_update)
+    all_dependent_ids = responses_to_update.map do |r|
```

Comment:
we can change this to:
```
all_dependent_ids = FrameworkResponse.where(id: responses_to_update.map(&:id)).includes(dependents: :framework_question).map do |response|
      response.dependents.reject { |dependent| response.option_selected?(dependent.framework_question.dependent_value) }
    end
```
This will avoid the n+1 queries

Reply:
Nice 👍